packages/worker/src/api/controllers/global/users.ts

```diff
@@ -84,12 +143,60 @@ export const adminUser = async (ctx: any) => {
 
 export const destroy = async (ctx: any) => {
   const id = ctx.params.id
+  const db = tenancy.getGlobalDB()
+  let user: User = await db.get(id)
+  let groups = user.userGroups
+
   await users.destroy(id, ctx.user)
+
+  // Remove asssosicated groups
+  if (groups) {
+    for (const groupId of groups) {
+      let group = await db.get(groupId)
+      let updatedUsersGroup = group.users.filter(
+        (groupUser: any) => groupUser.email !== user.email
+      )
+      group.users = updatedUsersGroup
+      await db.put(group)
+    }
+  }
+
   ctx.body = {
     message: `User ${id} deleted.`,
   }
 }
 
+export const bulkDelete = async (ctx: any) => {
+  const { userIds } = ctx.request.body
+  const db = tenancy.getGlobalDB()
+
+  let deleted = 0
+
+  for (const id of userIds) {
+    let user: User = await db.get(id)
+    let groups = user.userGroups
+
+    await users.destroy(id, ctx.user)
```

Comment:
Same as all above comments, this could be achieved with bulk operations, you can perform a bulk read of all the IDs initially (perform a lookup on the `allDocs` and use the `keys` attribute, e.g. `db.allDocs({ keys: userIds, include_docs: true })` this retrieves all of the users in a single speedy operation, then you can build up all the documents that need to be deleted, and attach a `_deleted: true` property on them during the `bulkDocs` call to delete them.

Ideally after seeing this, the logic for bulk saving and deletion of users could be extracted to the users core in the backend-core library, that way you could split out the synchronous portions of `users.destroy` and `users.save` to use in `users.bulkSave` and `users.bulkDestroy`.

Reply:
You mentioned backend-core here, is that right? Is there a reason for that, over putting the functionality into the users sdk within the worker? That seems to be where most of this type of functionality seems to go.